chains/heads/tracker.go

```diff
@@ -194,7 +200,28 @@ func (t *tracker[HTH, S, ID, BLOCK_HASH]) close() error {
 	return t.broadcastMB.Close()
 }
 
-func (t *tracker[HTH, S, ID, BLOCK_HASH]) Backfill(ctx context.Context, headWithChain HTH) (err error) {
+// verifyBlockHashes returns finality violated error if a block hash mismatch is found in provided chains
+func (t *tracker[HTH, S, ID, BLOCK_HASH]) verifyBlockHashes(headWithChain chains.Head[BLOCK_HASH], prevHeadWithChain chains.Head[BLOCK_HASH]) error {
+	if prevHeadWithChain == nil {
+		return nil
+	}
+
+	// Verify hashes from previous finalized chain until reaching the corresponding chain length
+	prevBlockNum := prevHeadWithChain.BlockNumber()
+	chainLength := int64(prevHeadWithChain.ChainLength())
+	for blockNum := prevBlockNum; blockNum > prevBlockNum-chainLength; blockNum-- {
```

Comment:
Do we need to iterate through the whole chain at all? It should be enough to verify if headWithChain contains a block at the same height as the latest finalized of prevHeadWithChain, that their hashes match.